### PR TITLE
Add Database Schema and Methods for Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR adds the database schema required by the optimized slasher and utilizes [eth2-types](https://github.com/prysmaticlabs/eth2-types) for greater type safety in the optimized slasher implementation. It includes the following:

- [x] Getter and setter for the latest epoch attested for validator indices, required for slashing detection
- [x] Getter and setter for attestation records for a validator indices at a given target epoch
- [x] Getter and setter for min/max span chunks for validator indices

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/db/kv/slasher.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	return s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationRecordsBucket)
+		encIdx, err := validatorIdx.MarshalSSZ()
+		if err != nil {
+			return err
+		}
+		encEpoch := ssz.MarshalUint64(make([]byte, 0), attestation.Data.Target.Epoch)
+		key := append(encIdx, encEpoch...)
+		value := make([]byte, 48)
+		copy(value[0:8], ssz.MarshalUint64(make([]byte, 0), attestation.Data.Source.Epoch))
```
Comment (terencechain) [inline beacon-chain/db/kv/slasher.go:118]: I think encode `IndexedAttestation` to `value` should be its own helper, and vice versa for decode

File: beacon-chain/db/kv/slasher.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		return bkt.Put(key, value)
+	})
+}
+
+// LoadSlasherChunk given a chunk kind and a disk key, retrieves a chunk for a validator
+// min or max span used by slasher from our database.
+func (s *Store) LoadSlasherChunk(
+	ctx context.Context, kind slashertypes.ChunkKind, diskKey uint64,
+) ([]uint16, bool, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.LoadChunk")
```
Comment (terencechain) [inline beacon-chain/db/kv/slasher.go:130]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "BeaconDB. LoadSlasherChunk")
```

File: beacon-chain/db/kv/slasher.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	})
+	return chunk, exists, err
+}
+
+// SaveSlasherChunk given a chunk kind, list of disk keys, and list of chunks,
+// saves the chunks to our database for use by slasher in slashing detection.
+func (s *Store) SaveSlasherChunks(
+	ctx context.Context, kind slashertypes.ChunkKind, chunkKeys []uint64, chunks [][]uint16,
+) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveChunks")
```
Comment (terencechain) [inline beacon-chain/db/kv/slasher.go:158]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "BeaconDB. SaveSlasherChunks")
```

File: beacon-chain/db/kv/slasher.go [hunk 4]
```diff
@@ -173,3 +173,29 @@ func (s *Store) SaveSlasherChunks(
 		return nil
 	})
 }
+
+// Encode an indexed attestation's required fields for slashing protection into bytes.
+func encodeIndexedAttestationRecord(att *ethpb.IndexedAttestation, signingRoot [32]byte) ([]byte, error) {
+	if att == nil || att.Data == nil || att.Data.Target == nil || att.Data.Source == nil {
```
Comment (terencechain) [inline beacon-chain/db/kv/slasher.go:179]: A good first issue is to add `ValidateNilIndexedAttestation` helper. Something to add when we are bored

**Review by terencechain — APPROVED** [review]
Lgtm

Had a minor concern on `SaveLatestEpochAttestedForValidators` being plural and `LatestEpochAttestedForValidator` being singular. This breaks symmetric and not sure what the performance impact is. Something to look at down the line

**Comment by rauljordan** [comment]
> Lgtm
> 
> Had a minor concern on `SaveLatestEpochAttestedForValidators` being plural and `LatestEpochAttestedForValidator` being singular. This breaks symmetric and not sure what the performance impact is. Something to look at down the line

Makes sense, we will revisit this once the functions are used at runtime to evaluate if we need these batched methods

**Pull request merged**